lib/translate/status_translator.cc

```diff
@@ -0,0 +1,240 @@
+          scsi_defs::AdditionalSenseCode::kLbaOutOfRange};
+      break;
+    case nvme_defs::GenericCommandStatusCode::kNamespaceNotReady:
+      status = scsi_defs::Status::kCheckCondition;
+      sense_key = scsi_defs::SenseKey::kNotReady;
+      sense_code = std::optional<scsi_defs::AdditionalSenseCode>{
+          scsi_defs::AdditionalSenseCode::
+              kLogicalUnitNotReadyCauseNotReportable};
+      break;
+    default:
+      printf("No SCSI translation for NVMe status with code %#x\n",
+             (uint8_t)sc);
+      return nullptr;
+  }
+  return new TranslatedScsiStatus(status, sense_key, sense_code);
+}
+
+// Section 7.2
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+TranslatedScsiStatus *translateStatusCode(
+    const nvme_defs::CommandSpecificStatusCode sc) {
+  scsi_defs::Status status;
+  scsi_defs::SenseKey sense_key;
+  std::optional<scsi_defs::AdditionalSenseCode> sense_code = std::nullopt;
+
+  switch (sc) {
+    case nvme_defs::CommandSpecificStatusCode::kCompletionQueueInvalid:
+      status = scsi_defs::Status::kCheckCondition;
+      sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      break;
+    case nvme_defs::CommandSpecificStatusCode::
+        kInvalidQueueIdentifier:  // unspecified
+    case nvme_defs::CommandSpecificStatusCode::
+        kInvalidQueueSize:  // unspecified
+    case nvme_defs::CommandSpecificStatusCode::
+        kAbortCommandLimitExceeded:  // unspecified
+    case nvme_defs::CommandSpecificStatusCode::
+        kAsyncEventRequestLimitExceeded:  // unspecified
+    case nvme_defs::CommandSpecificStatusCode::
+        kInvalidFirmwareSlot:  // unspecified
+    case nvme_defs::CommandSpecificStatusCode::
+        kInvalidFirmwareImage:  // unspecified
+    case nvme_defs::CommandSpecificStatusCode::
+        kInvalidInterruptVector:                                 // unspecified
+    case nvme_defs::CommandSpecificStatusCode::kInvalidLogPage:  // unspecified
+      status = scsi_defs::Status::kCheckCondition;
+      sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      break;
+    case nvme_defs::CommandSpecificStatusCode::kInvalidFormat:
+      status = scsi_defs::Status::kCheckCondition;
+      sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      sense_code = std::optional<scsi_defs::AdditionalSenseCode>{
+          scsi_defs::AdditionalSenseCode::kFormatCommandFailed};
+      break;
+    case nvme_defs::CommandSpecificStatusCode::kConflictingAttributes:
+      status = scsi_defs::Status::kCheckCondition;
+      sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      sense_code = std::optional<scsi_defs::AdditionalSenseCode>{
+          scsi_defs::AdditionalSenseCode::kInvalidFieldInCdb};
+      break;
+    default:
+      printf("No SCSI translation for NVMe status with code %#x\n",
+             (uint8_t)sc);
+      return nullptr;
+  }
+  return new TranslatedScsiStatus(status, sense_key, sense_code);
```

Comment:
Given recent memory discussions we may want to offload this allocation to outside of this file and have a pointer to the allocated struct as a parameter to the translation method

Reply:
good point!